medusa/tv/episode.py

```diff
@@ -87,20 +84,209 @@
     get_scene_numbering,
     xem_refresh,
 )
-from medusa.tv.base import TV
-
-import shutil_custom
+from medusa.tv.base import Identifier, TV
 
 try:
     import xml.etree.cElementTree as ETree
 except ImportError:
     import xml.etree.ElementTree as ETree
 
-shutil.copyfile = shutil_custom.copyfile_custom
+logger = logging.getLogger(__name__)
 
-MILLIS_YEAR_1900 = datetime.datetime(year=1900, month=1, day=1).toordinal()
 
-logger = logging.getLogger(__name__)
+class InvalidNumbering(AssertionError):
+    """Numbering is invalid"""
+
+
+class EpisodeNumber(object):
+    """Base identifier class."""
+
+    _attrs = []
+
+    def __eq__(self, other):
+        """Compare episodes for equality."""
+        others = (
+            getattr(other, attr, NotImplemented)
+            for attr in self._attrs
+        )
+        return all(x==y for x,y in zip(self, others))
+
+    def __ne__(self, other):
+        """Check episodes for inequality."""
+        return not self == other
+
+    def __iter__(self):
+        for attr in self._attrs:
+            yield getattr(self, attr)
+
+    def __bool__(self):
+        """Check boolean value of episode numbering."""
+        return all(self)
+
+    __nonzero__ = __bool__
+
+    def __hash__(self):
+        """Use attributes to generate a hash."""
+        return hash(tuple(self))
+
+    def __str__(self):
+        """Convert to string."""
+        return ', '.join(
+            '{0}={1!r}'.format(
+                attr, getattr(self, attr)
+            )
+            for attr in self._attrs
+            if getattr(self, attr) != NotImplemented
+        )
+
+    def __repr__(self):
+        """Class representation."""
+        return '<{self.__class__.__name__} [{self!s}]>'.format(self=self)
+
+    @classmethod
+    def parse(cls, number):
+        raise NotImplementedError
+
+    @classmethod
+    def from_string(cls, numbering):
+        return cls(**cls.parse(numbering))
+
+
+class AbsoluteNumber(EpisodeNumber):
+    """Season-less episode number."""
+
+    _attrs = ['episode']
+
+    regex = re.compile(
+        r'(?:s(?P<season>\d{1,4}))?(?:e?(?P<episode>\d+))',
```

Comment:
The idea here is to minimize the number of regex's required to reduce parsing.  The parent class doesn't really know anything about its child class, but it does know that Seasons exist and that it should not have any.